### PR TITLE
error budgets: add SLOs and SLIs

### PR DESCRIPTION
**Pull request by TomasTomecek**
TODO
* [x] review CKIs SLOs
* [x] review TFs SLOs and SLIs
* [x] discuss with the team
* [x] research build duration implementation with prometheus

### CODE REVIEW

File: error-budgets/SLOs-SLIs.md [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+and a pretty slick
+[dashboard](http://grafana-latest-osci-infra.cloud.privileged.psi.redhat.com/d/VXwH27XMk/rhel-infrastructure-and-service-health?orgId=1).
+
+Cockpit also uses this framework and the team has [a nice dashboard as well](http://grafana-latest-osci-infra.cloud.privileged.psi.redhat.com/d/VgtdtiwGk/rhel-infrastructure-and-service-health?orgId=1).
+
+## Packit SLOs
+
+successful run = job finished and everything is awesome
+failure = the job finished and the result is failure: the user needs to fix it
+error = infrastructure problems prevented the job to complete
```
Comment (jpopelka) [inline error-budgets/SLOs-SLIs.md:23]: They render on one line
```suggestion
- successful run = job finished and everything is awesome
- failure = the job finished and the result is a failure: the user needs to fix it
- error = infrastructure problems prevented the job to complete
```

File: error-budgets/SLOs-SLIs.md [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+[SLOs](https://cki-project.org/docs/hacking/rfcs/cki-004-slo/#created-slos-based-on-the-feedback).
+Instead of taking the same route, let's start small and slowly add more and
+more objectives and indicators over time.
+
+Testing Farm team has [Service level
+expectations](https://docs.google.com/document/d/1iHK3v_tCK4w-F82ZvUEO3xHcNv78tkR31w3oF-zpi1Q/edit#heading=h.26gd46u8e15q)
+and a pretty slick
+[dashboard](http://grafana-latest-osci-infra.cloud.privileged.psi.redhat.com/d/VXwH27XMk/rhel-infrastructure-and-service-health?orgId=1).
+
+Cockpit also uses this framework and the team has [a nice dashboard as well](http://grafana-latest-osci-infra.cloud.privileged.psi.redhat.com/d/VgtdtiwGk/rhel-infrastructure-and-service-health?orgId=1).
```
Comment (jpopelka) [inline error-budgets/SLOs-SLIs.md:17]: "Not found"

File: error-budgets/SLOs-SLIs.md [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+
+## Packit SLOs
```
Comment (jpopelka) [inline error-budgets/SLOs-SLIs.md:19]: Can we have a link to the [Users expectations](https://github.com/packit/research/blob/main/error-budgets/users-expectations.md) somewhere here?

File: error-budgets/SLOs-SLIs.md (hunk 2 above)
Comment (TomasTomecek) [reply error-budgets/SLOs-SLIs.md:17]: you're right, asked the team about this

File: error-budgets/SLOs-SLIs.md [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+
+SLOs are based on [discussions with our stakeholders](users-expectations.md).
+
+Let's define job statuses first to get a clear understanding of these objectives:
+
+- successful run: a job finished and everything is awesome
+- failure: the job finished and wasn't successful: the PR needs to be fixed
+- error: infrastructure problems prevented the job to complete
+
+### 99% of builds finish with success/failure by 24 hours
```
Comment (csomh) [inline error-budgets/SLOs-SLIs.md:29]:
```suggestion
### 99% of builds finish with success or failure within 24 hours
```

... but this seem to somewhat contradict the practice to babysit copr build for about [8 hours](https://github.com/packit/packit-service/blob/ff58a490b6e78a57e6f5655b063ef3ccf4c6addf/packit_service/worker/tasks.py#L73). Maybe 8 hours is too little, but 24 hours seems too long, if I think about this in the context of CI. Worst case scenario, I would expect 12 hours. Anything above this seems to render any result irrelevant. But this is all speculation, probably would make sense to ask some prominent users, what do they expect. What is the maximum time for a build to finish, which is still useful for them and doesn't slow down their development too much.

File: error-budgets/SLOs-SLIs.md [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+
+The most core functionality is to run COPR builds for PRs. We want to be sure
+those builds either pass or fail and no error interrupts the build process.
+
+The problem is that some builds take minutes and some hours so it's hard to
+design this objective to make it more proactive - that we'd know sooner that
+something is wrong.
```
Comment (csomh) [inline error-budgets/SLOs-SLIs.md:36]: I don't think there is anything wrong with that. I don't expect this objective to be used for short term monitoring.

File: error-budgets/SLOs-SLIs.md [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+
+### 95% of test runs finish with success/failure by 24 hours
```
Comment (csomh) [inline error-budgets/SLOs-SLIs.md:38]:
```suggestion
### 95% of test runs finish with success or failure within 24 hours
```

The same concerns I had regarding the duration above.

File: error-budgets/SLOs-SLIs.md [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+
+- Number of builds that finished with an error
+
+- Number of unfinished builds that are in progress for more than 24 hours
+
+- Number of test runs that finished with an error
+
+- Number of PRs handled by packit with no commit status from packit in >15s
+
+- Time it takes packit to set the initial status
```
Comment (csomh) [inline error-budgets/SLOs-SLIs.md:60]: These indicators seem to me to be derivatives.

Shouldn't we have some more basic ones which can be combined to derive and measure the objectives? Like: number of builds started, number of builds finished with successful/failure/error, build duration, etc...

File: error-budgets/SLOs-SLIs.md (hunk 4 above)
Comment (TomasTomecek) [reply error-budgets/SLOs-SLIs.md:29]: The problem here is that every project is different and builds take different times for ogr, cockpit, anaconda, kernel, systemd, atlas, gcc... One absolute number can't fit all. 

The indicator I'd like to see here would measure [moving average (MA)](https://en.wikipedia.org/wiki/Moving_average) and the object would require have consistent. Not sure how hard would it be to [measure MA](https://stackoverflow.com/questions/27489056/is-there-a-way-to-have-a-moving-average-in-grafana) with grafana.

**Review by csomh — APPROVED** [review]

**Pull request merged**